### DISCUSSION TITLE
Can pymilvus be used with Flask or Jango? What about concurrency (several requests at once)?

### DISCUSSION THREAD

**Discussion by martinenkoEduard** [milvus-io/milvus#18194] [Q&A and General discussion]
Can pymilvus be used with Flask or Jango? What about concurrency (several requests at once)?

**Comment by yhmo** [comment]
Flask and Jango are web framework. Pymilvus is a lib. Definitely pymilvus can be used with Flask or Jango.
Concurrency performance is mainly depends on several aspects: the hardware, deployment, dataset size, index type/parameters, search parameters, etc.

**Reply by martinenkoEduard** [reply]
I mean where should I put connection string?
`connections.connect("default", host="localhost", port="19530")`


What is the right way? Considering Gunicorn ..




```
from flask import Flask
app = Flask(__name__)
connections.connect("default", host="localhost", port="19530")

@app.route('/')
def hello():
    return 'Hello, World!'
```


or

```
from flask import Flask
app = Flask(__name__)
@app.route('/')
def hello():
    connections.connect("default", host="localhost", port="19530")
    return 'Hello, World!'
```

**Reply by martinenkoEduard** [reply]
I mean should I create a new connection each time or I can use one connection? 
Does it work well with gunicorn?

**Reply by yhmo** [reply]
The `connections.connect("default", host="localhost", port="19530")` create a connection with a name, the "default" is the name of the connection. The **connections** object maintains a map of grpc connections, key is the name, value is a connection.
When you call `connections.connect("default", host="localhost", port="19530")`, a connection is created with name "default". 

Assume you have a collection, now you want to do something to the collection, firstly you declare a Collection object:
`coll = Collection(name="aaa", using="default")`
The "using" parameter allows you to specify which connection should be used, here we input "default", which means this object will use the connection named "default".

Then you do something to the "coll" object:
`coll.load()`
This line will call the load() API by the "default" connection.

If you don't need the connection anymore, you can call `connections.remove_connection("default")` to close the connection.

In my opinion, both two ways can work, but are not efficient:
1. create one connection for all requests
2. create a new connection for each request

It is better to build some kind of connection pool, prepare some connections during initializing, and categorize your requests into several kinds, each kind uses a connection.
